File: 2015/Day 15/day15.cpp
//
//  Advent of Code 2015
//  Day 01
//
//  By PatchesTheDipstick
//


//  System includes...


//  Third Party includes...



//  Project includes...

#include "../utils.hpp"

struct Ingredient
{
	std::string name;
	int capacity;
	int durability;
	int flavor;
	int texture;
	int calories;
};

int calculate(std::vector<Ingredient> stock, int a, int b, int c, int d)
{
	int return_value = 0;
	int cap = stock[0].capacity * a;
	cap += stock[1].capacity * b;
	cap += stock[2].capacity * c;
	cap += stock[3].capacity * d;
	//std::cout << "cap " << cap << std::endl;
	if (cap < 0)
	{
		cap = 0;
	}
	int dur = stock[0].durability * a;
	dur += stock[1].durability * b;
	dur += stock[2].durability * c;
	dur += stock[3].durability * d;
	//std::cout << "dur " << dur << std::endl;
	if (dur < 0)
	{
		dur = 0;
	}
	int fla = stock[0].flavor * a;
	fla += stock[1].flavor * b;
	fla += stock[2].flavor * c;
	fla += stock[3].flavor * d;
	//std::cout << "fla " << fla << std::endl;
	if (fla < 0)
	{
		fla = 0;
	}
	int tex = stock[0].texture * a;
	tex += stock[1].texture * b;
	tex += stock[2].texture * c;
	tex += stock[3].texture * d;
	//std::cout << "tex " << tex << std::endl;
	if (tex < 0)
	{
		tex = 0;
	}
	
	return_value = cap * dur * fla * tex;
	return return_value;
}

int countCalories(std::vector<Ingredient> stock, int a, int b, int c, int d)
{
	int return_value = 0;
	int cal = stock[0].calories * a;
	cal += stock[1].calories * b;
	cal += stock[2].calories * c;
	cal += stock[3].calories * d;
	return_value = cal;
	return return_value;
}

int main(int argc, char** argv)
{
	
	std::vector<Ingredient> cupboard = {{"Frosting", 4, -2, 0, 0, 5},
										{"Candy", 0, 5, -1, 0, 8},
										{"Butterscotch", -1, 0, 5, 0, 6},
										{"Sugar", 0, 0, -2, 2, 1}};
	

	std::vector<int> results;
	std::vector<int> results2;
	for (int a = 0; a <= 100; a++)
	{
		for (int b = 0; b <= 100; b++)
		{
			for (int c = 0; c <= 100; c++)
			{
				for (int d = 0; d <= 100; d++)
				{
					if (a + b + c + d == 100)
					{
						int temp = calculate(cupboard, a, b, c, d);
						if (temp > 0)
						{
							results.push_back(temp);
							if (countCalories(cupboard, a, b, c, d) == 500)
							{
								results2.push_back(temp);
							}
						}
					}
				}
			}
		}
	}
	int answer_1 = 0;
	
	for (int a = 0; a < results.size(); a++)
	{
		if (results[a] > answer_1)
		{
			answer_1 = results[a];
		}
	}
	std::cout << "Advent of Code 2015 Day 1 answers:" << std::endl;
	std::cout << "    Part 1 : " << answer_1 << std::endl;
	
	int answer_2 = 0;
	std::cout << results2.size() << std::endl;
	for (int a = 0; a < results2.size(); a++)
	{
		if (results2[a] > answer_2)
		{
			answer_2 = results2[a];
		}
	}
	
	std::cout << "    Part 2 : " << answer_2 << std::endl;

	return (0);

}

File: 2015/Day 15/day15.py
#
#  Advent of Code 2015
#  Day 
#
#  By PatchesTheDipstick
#
import re

def getInput(file_name):
	input = []
	with open(file_name) as input_file:
		input = input_file.read().splitlines()
	return input

def get_score(recipe, ingredient):

	score = 1
	for j in range(4):
		sub = 0
		for i, k in enumerate(recipe):
			sub += ingredient[i][j] * k
		if sub < 0:
			sub = 0
		score *= sub
	return score

def generate_recipes():
	recipes = []
	for a in range(101):
		for b in range(101):
			for c in range(101):
				for d in range(101):
					if a + b + c + d == 100:
						recipes.append((a, b, c, d))
	return recipes

def get_calories(ingredient, recipe):
	c = 0
	for i in range(4):
		c += ingredient[i][4] * recipe[i]
	return c

def part_one(ingredient, recipes):
	score = 0
	for recipe in recipes:
		s = get_score(recipe, ingredient)
		if s > score:
			score = s
	return score

def part_two(ingredient, recipes):
	score = 0
	for recipe in recipes:
		if get_calories(ingredient, recipe) == 500:
			s = get_score(recipe, ingredient)
			if s > score:
				score = s
	return score

if __name__ == '__main__':
	
	ingredient = {}
	input = getInput("input.txt")
	rule = "^(.+): capacity (.+), durability (.+), flavor (.+), texture (.+), calories (.+)$"
	for i, line in enumerate(input):
		token = re.search(rule, line)
		ingredient[i] = (int(token.group(2)), int(token.group(3)), int(token.group(4)), int(token.group(5)), int(token.group(6)), token.group(1))
	recipes = generate_recipes()

	print("Advent of Code 2015 Day 03")
	print("  Part one:", part_one(ingredient, recipes))
	print("  Part two:", part_two(ingredient, recipes))	
	